src/Libraries/RevitNodes/Application/Document.cs

```diff
@@ -62,6 +66,171 @@ public static Document Current
+        {
+            //The internal GUID of the purge Performance Adviser Rule 
+            const string purgePerformanceAdviserRuleGuid = "e8c63650-70b7-435a-9010-ec97660c1bda";
+
+            IList<PerformanceAdviserRuleId> performanceAdviserRules = PerformanceAdviser.GetPerformanceAdviser().GetAllRuleIds();
+            List<PerformanceAdviserRuleId> performanceAdviserRuleId = performanceAdviserRules.Where(x => x.Guid.ToString() == purgePerformanceAdviserRuleGuid).ToList();
+            List<ElementId> purgedElementIds = new List<ElementId>();
+            TransactionManager.Instance.EnsureInTransaction(this.InternalDocument);
+            
+            purgedElementIds.AddRange(PurgeElements(performanceAdviserRuleId, deepPurge));
+            purgedElementIds.AddRange(PurgeMaterials());
+
+            TransactionManager.Instance.TransactionTaskDone();
+
+            if (purgedElementIds.Count == 0)
+                throw new InvalidOperationException(Properties.Resources.NoElementsToPurge);
+
+            return purgedElementIds.Select(x => x.IntegerValue).ToList();
+        }
+
+        private List<ElementId> PurgeMaterials()
+        {
+            List<Autodesk.Revit.DB.ElementId> materials = new FilteredElementCollector(this.InternalDocument).OfClass(typeof(Autodesk.Revit.DB.Material))
+                                                                                                           .ToElementIds()
+                                                                                                           .ToList();
+
+            List<Autodesk.Revit.DB.Element> elementTypes = new FilteredElementCollector(this.InternalDocument).WhereElementIsElementType()
+                                                                                                              .ToElements()
+                                                                                                              .ToList();
+            var nonPurgedElements = new List<Autodesk.Revit.DB.ElementId>();
+
+            for (int i = 0; i < elementTypes.Count; i++)
+            {
+                HostObjAttributes hostObj = elementTypes[i] as HostObjAttributes;
+                if (hostObj != null)
+                {
+                    var compundStructure = hostObj.GetCompoundStructure();
+                    if (compundStructure == null)
+                        continue;
+
+                    int layerCount = compundStructure.LayerCount;
+                    for (int j = 0; j < layerCount; j++)
+                    {
+                        ElementId elementId = compundStructure.GetMaterialId(j);
+                        if (materials.Contains(elementId))
+                        {
+                            nonPurgedElements.Add(elementId);
+                            materials.Remove(elementId);
+                        }
+                        
+                    }
+                    continue;
```

Comment:
this should be moved to a `private static` utility method that returns all material ids for an `ElementType` - the cast to `HostObjAttributes` etc can be hidden in that method

this means the method will return a list of element ids and you'll need to affect the `materials` and `nonPurged` list outside of that method